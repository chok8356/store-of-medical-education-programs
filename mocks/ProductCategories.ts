import { ICategory } from '../types/ICategory'

export const PRODUCTS_CATEGORIES_COUNT = 5

export const categories: ICategory[] = new Array(PRODUCTS_CATEGORIES_COUNT)
  .fill(null)
  .map((x, i) => {
    const id = i + 1
    return {
      id,
      name: `Категория ${id}`
    }
  })
