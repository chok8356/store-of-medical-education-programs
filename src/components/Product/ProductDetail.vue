<template>
  <b-card
    v-if="product"
    class="overflow-hidden">
    <b-row>
      <b-col md="6">
        <b-card-img :src="product.img" />
      </b-col>
      <b-col md="6">
        <b-card-body>
          <h1 class="mb-4">
            {{ product.name }}
          </h1>

          <b-card-text>
            {{ product.description }}
          </b-card-text>

          <b-badge variant="info">
            {{ category.name }}
          </b-badge>

          <div class="d-block mt-4">
            <b-button
              variant="primary"
              @click="addToCart">
              В корзину
            </b-button>
          </div>
        </b-card-body>
      </b-col>
    </b-row>
  </b-card>
</template>

<script lang="ts">
import { BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop } from 'vue-property-decorator'
import { ICategory } from '../../../types/ICategory'
import { IProduct } from '../../../types/IProduct'

@Component({
  name: 'ProductDetail',
  components: {
    BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg
  }
})
export default class ProductDetail extends Vue {
  @Prop({ type: Object as PropType<IProduct> }) product!: IProduct
  @Prop({ type: Object as PropType<ICategory> }) category!: ICategory

  addToCart() {
    this.$emit('add-to-cart')
  }
}
</script>
