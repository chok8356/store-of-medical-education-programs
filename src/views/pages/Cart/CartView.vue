<template>
  <div class="cart-view">
    <b-button
      class="mb-4"
      @click="backToStore">
      Назад в магазин
    </b-button>

    <h4>Товары</h4>
    <b-table
      v-loading="cartLoading"
      borderless
      :fields="fields"
      fixed
      hover
      :items="items"
      show-empty
      striped>
      <template #cell(price)="data">
        {{ data.value | currency }}
      </template>

      <template #cell(actions)="data">
        <b-icon
          class="cart-view__table-icon mr-1"
          icon="dash-square"
          @click="onRemoveClick(data.item)" />
        <b-icon
          class="cart-view__table-icon"
          icon="plus-square"
          @click="onAddClick(data.item)" />
      </template>

      <template
        v-if="items.length"
        #custom-foot>
        <b-tr>
          <b-th>
            Всего:
          </b-th>
          <b-th>
            {{ cartTotal | currency }}
          </b-th>
          <b-th>
            {{ cartQuantity }}
          </b-th>
        </b-tr>
      </template>

      <template #empty>
        Нет товаров
      </template>
    </b-table>

    <b-button
      v-if="items.length"
      class="mt-4"
      @click="clearCart">
      Очистить корзину
    </b-button>
  </div>
</template>

<script lang="ts">
import { BButton, BTable, BTr, BTh, BIcon } from 'bootstrap-vue'
import { Vue, Component } from 'vue-property-decorator'
import { namespace } from 'vuex-class'
import { loading } from '@/directives'
import { currency } from '@/filters'
import { ICart, ICartItem } from '../../../../types/ICart'

const CartStore = namespace('cart')

@Component({
  name: 'CartView',
  directives: {
    loading
  },
  filters: {
    currency
  },
  components: {
    BButton, BTable, BTr, BTh, BIcon
  }
})
export default class CartView extends Vue {
  @CartStore.State cart!: ICart
  @CartStore.State cartLoading!: boolean

  @CartStore.Getter('total') cartTotal!: number
  @CartStore.Getter('quantity') cartQuantity!: number

  @CartStore.Action addToCart!: any
  @CartStore.Action removeFromCart!: any
  @CartStore.Action clearCart!: any

  fields = ['name', 'price', 'quantity', 'actions']

  get items() {
    return Object.values(this.cart)
  }

  backToStore() {
    this.$router.push({ name: 'ProductsList' })
  }

  onAddClick(product: ICartItem) {
    this.addToCart(product.id)
  }

  onRemoveClick(product: ICartItem) {
    this.removeFromCart(product.id)
  }
}
</script>

<style lang="scss">
.cart-view {
  &__table-icon {
    cursor: pointer;

    &:hover {
      color: var(--blue);
    }
  }
}
</style>
