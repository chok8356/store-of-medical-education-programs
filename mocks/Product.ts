import { IProduct } from '../types/IProduct'
import { PRODUCTS_CATEGORIES_COUNT } from './ProductCategories'

const PRODUCTS_COUNT = 100

export const products: IProduct[] = new Array(PRODUCTS_COUNT).fill(null).map((x, i) => {
  const id = i + 1
  return {
    id,
    name: `Продукт ${id}`,
    price: id + PRODUCTS_COUNT,
    // eslint-disable-next-line max-len
    img: 'https://media.istockphoto.com/photos/stack-of-medical-student-textbooks-picture-id921644532?k=20&m=921644532&s=612x612&w=0&h=hezQpZ0uyUmbU4QJePKK5JPNWVXD_UWwWPYSg5Rspds=',
    // eslint-disable-next-line max-len
    description: id % 7 + ' Полностью новые главы по биохимии и клеточной биологии, генетике, нервной системе, костям, мышцам и коже, эндокринной и репродуктивной системам, сердечно-сосудистой системе, почечной системе, диете и питанию.',
    category: id % PRODUCTS_CATEGORIES_COUNT + 1
  }
})
