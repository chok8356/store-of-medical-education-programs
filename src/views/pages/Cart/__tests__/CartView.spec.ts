import { createLocalVue, shallowMount } from '@vue/test-utils'

import { BButton } from 'bootstrap-vue'
import Vuex from 'vuex'
import { products } from '@/../mocks/Product'

import CartStore from '@/store/cart'
import CartView from '../CartView.vue'

const product = products[0]

const localVue = createLocalVue()
localVue.use(Vuex)

describe('CartView', () => {
  let wrapper
  let actions
  let store

  beforeEach(() => {
    actions = {
      addToCart: jest.fn(),
      removeFromCart: jest.fn(),
      clearCart: jest.fn()
    }
    store = {
      modules: {
        cart: {
          ...CartStore,
          state: {
            cart: {
              [product.id]: product
            }
          },
          actions
        }
      }
    }
    wrapper = shallowMount(CartView, {
      store: new Vuex.Store(store),
      localVue,
      mocks: {
        $router: { push: () => {} }
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('render correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('computed items', () => {
    expect(wrapper.vm.items).toStrictEqual([product])
  })

  it('back to store', () => {
    const routerPush = jest.spyOn(wrapper.vm.$router, 'push')
    const buttonComponent = wrapper.findComponent(BButton)
    buttonComponent.trigger('click')
    expect(routerPush).toBeCalled()
  })

  it('clear cart', async() => {
    const buttonComponent = wrapper.findAllComponents(BButton).at(1)
    buttonComponent.trigger('click')
    expect(actions.clearCart).toHaveBeenCalled()
  })

  it('add to cart', async() => {
    wrapper.vm.onAddClick(product)
    expect(actions.addToCart).toHaveBeenCalled()
  })

  it('remove from cart', async() => {
    wrapper.vm.onRemoveClick(product)
    expect(actions.removeFromCart).toHaveBeenCalled()
  })
})
