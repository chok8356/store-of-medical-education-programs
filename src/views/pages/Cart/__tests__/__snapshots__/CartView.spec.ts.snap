// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartView render correctly 1`] = `
<div
  class="cart-view"
>
  <b-button-stub
    class="mb-4"
    tag="button"
    target="_self"
    type="button"
    variant="secondary"
  >
    
    Назад в магазин
  
  </b-button-stub>
   
  <h4>
    Товары
  </h4>
   
  <b-table-stub
    borderless="true"
    currentpage="1"
    emptyfilteredtext="There are no records matching your request"
    emptytext="There are no records to show"
    fields="name,price,quantity,actions"
    filterdebounce="0"
    filterfunction="[Function]"
    filterignoredfields=""
    filterincludedfields=""
    fixed="true"
    hover="true"
    items="[object Object]"
    labelsortasc="Click to sort Ascending"
    labelsortclear="Click to clear sorting"
    labelsortdesc="Click to sort Descending"
    perpage="0"
    selectedvariant="active"
    selectmode="multi"
    showempty="true"
    sortcompare="[Function]"
    sortcompareoptions="[object Object]"
    sortdirection="asc"
    striped="true"
    value=""
  />
   
  <b-button-stub
    class="mt-4"
    tag="button"
    target="_self"
    type="button"
    variant="secondary"
  >
    
    Очистить корзину
  
  </b-button-stub>
</div>
`;
